Velbus auto-discovery

thomasdelaet opened the pull request:
## Description:

This pull requests simplifies velbus platforms by moving most of the logic to python-velbus. It now supports auto-discovery of modules.

**Breaking change**
Breaking changes include auto-discovery support so that you don't need to specify modules manually anymore in the configuration file, but just load the velbus component which will automatically load the switch and binary_sensor platforms. In addition, the fan and light platforms are removed since these should be modeled with template platform based on velbus switches. 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5132

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

thomasdelaet committed:
remove velbus fan and light platforms

these platforms should not be there since they can be created with template components based on switch platform

thomasdelaet committed:
use latest version of python-velbus which supports auto-discovery of modules

thomasdelaet committed:
fix linting errors

thomasdelaet committed:
fix linting errors

thomasdelaet commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +28,25 @@
+    def callback():
+        hass.async_add_job(async_load_platform(hass, 'switch', DOMAIN))
+        hass.async_add_job(async_load_platform(hass, 'binary_sensor', DOMAIN))
+    
```
fixed

thomasdelaet commented on homeassistant/components/switch/velbus.py:
```diff
@@ -8,73 +8,60 @@
+    modules = velbus.get_modules('switch')
+    for module in modules:
+        for channel in range(1, module.number_of_channels() + 1):
+            async_add_devices([VelbusSwitch(module, channel)], update_before_add=True)
```
fixed

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,60 @@
+DEPENDENCIES = ['velbus']
 
-def setup_platform(hass, config, add_devices, discovery_info=None):
+@asyncio.coroutine
```
fixed

thomasdelaet committed:
fix linting errors

thomasdelaet commented:
Happy to provide more info on this one if somebody requires it. Would be great to get some feedback on what is required to get this merged.

thomasdelaet commented:
That would be great! @pvizeli : this is the change you suggested back then around platform discovery.

MartinHjelmare commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,62 @@
+
+        def callback():
+            """Callback when module is loaded."""
+            future.set_result(None)
```
I think a future's result should be set from within the event loop, so schedule it with `hass.add_job`.

thomasdelaet replied:
ok

MartinHjelmare commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +28,26 @@
+
+    def callback():
+        """Callback when scan is completed."""
+        hass.async_add_job(async_load_platform(hass, 'switch', DOMAIN))
```
Add an empty dict as discovery_info (fourth argument) and the original config as fifth argument.

thomasdelaet replied:
ok. See also other comment: if discovery_info is empty, why should I return on discovery_info being None in the platforms?

MartinHjelmare commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +28,26 @@
+        hass.async_add_job(async_load_platform(hass, 'switch', DOMAIN))
+        hass.async_add_job(async_load_platform(hass, 'binary_sensor', DOMAIN))
+
+    controller.scan(callback)
```
This call contains a `time.sleep` (3 seconds) call. Sleeping is not allowed inside the event loop. Doing it in a worker thread is not optimal either, since if that would be done frequently, the thread pool would be exhausted. Is there another option than sleeping?

thomasdelaet replied:
@MartinHjelmare Thank you for your comments. I'm working on incorporating these in the PR.

On the timer: time.sleep is not executed in the event loop but in a thread (see __init__.py in connections directory: the message and callback is put in a write queue that is read by the write thread.

I don't see any alternative to a sleep command given that there is no way to know which addresses on the bus are being used, hence I iterate over all addresses, wait for a couple of seconds and see which ones respond and which ones don't. Happy to get feedback if you would see an alternative solution.

MartinHjelmare replied:
Sleeping in a separate thread is fine. :+1:

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,62 @@
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up Velbus binary sensors."""
     velbus = hass.data[DOMAIN]
```
@MartinHjelmare why would I do that? I'm not getting the sensors from discovery_info object and this will always be empty, or am I missing something? Thanks for clarifying.

MartinHjelmare replied:
Because we only want to set up the platform via discovery from the component and not via platform config. `discovery_info` will be `None` if platform config is used to set up the platform.

MartinHjelmare replied:
It's ok to keep it like it's done now. But if you see an opportunity for simplification, go ahead.

MartinHjelmare replied:
`discovery_info` must be serializable though.

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,62 @@
 For more details about this platform, please refer to the documentation at
 https://home-assistant.io/components/binary_sensor.velbus/
 """
+
```
ok

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,62 @@
 
 
-def setup_platform(hass, config, add_devices, discovery_info=None):
+@asyncio.coroutine
```
ok

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,62 @@
+    for module in modules:
+        for channel in range(1, module.number_of_channels() + 1):
+            sensor = VelbusBinarySensor(module, channel)
+            async_add_devices([sensor], update_before_add=True)
```
ok

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,62 @@
+        def _init_velbus():
+            """Initialize Velbus on startup."""
+            self._module.on_status_update(self._channel, self._on_update)
+        yield from self.hass.async_add_job(_init_velbus)
```
ok

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,62 @@
+        for channel in range(1, module.number_of_channels() + 1):
+            sensor = VelbusBinarySensor(module, channel)
+            async_add_devices([sensor], update_before_add=True)
+    return True
```
ok

thomasdelaet committed:
address review comments from @MartinHjelmare

thomasdelaet committed:
update based on automatic feedback

thomasdelaet committed:
fix linting errors

thomasdelaet committed:
update dependency

thomasdelaet committed:
syntax corrections

thomasdelaet committed:
fix lint warning

MartinHjelmare commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +27,42 @@
+                            module.get_module_address(),
+                            channel
+                        ))
+        hass.async_add_job(async_load_platform(hass, 'switch', DOMAIN,
```
Since the callback will be called from a separate thread, we can't use async methods here. Change to use `load_platform` directly. It's not needed to add a job for that.

MartinHjelmare commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,82 +4,58 @@
     """Representation of a Velbus Binary Sensor."""
 
-    def __init__(self, binary_sensor, velbus):
+    def __init__(self, module, channel):
```
Break out the common methods and properties into a common base velbus device entity class defined in the velbus component. Then use multiple inheritance in the platform entities, when defining those classes.

thomasdelaet committed:
split out common functionality in VelbusEntity
use sync methods for loading platforms
support unique_ids so that entities are registred in entity registry

thomasdelaet committed:
fix linting errors

thomasdelaet committed:
fix linting errors

thomasdelaet committed:
fix linting errors

thomasdelaet commented:
@MartinHjelmare changed based on your latest comments. Could you take another look and merge if okay for you?

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,93 +4,68 @@
+        self._module.on_status_update(self._channel, callback)
+
+    def _on_update(self, state):
+        self.schedule_update_ha_state()
```
Done

thomasdelaet commented on homeassistant/components/binary_sensor/velbus.py:
```diff
@@ -4,93 +4,60 @@
+
+    async def async_update(self):
+        """Update module status."""
+        await self._load_module()
```
Done

thomasdelaet commented on homeassistant/components/switch/velbus.py:
```diff
@@ -84,28 +47,20 @@ def should_poll(self):
+
+    async def async_update(self):
+        """Update module status."""
+        await self._load_module()
```
Done

thomasdelaet commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +27,66 @@
+            future.set_result(None)
+
+        self._module.load(callback)
+        return future
```
Done

MartinHjelmare commented on homeassistant/components/switch/velbus.py:
```diff
@@ -4,77 +4,41 @@
+        return self._module.get_name(self._channel)
 
     @property
     def should_poll(self):
```
Move this too.

MartinHjelmare commented on homeassistant/components/switch/velbus.py:
```diff
@@ -84,28 +48,27 @@ def should_poll(self):
-        self._velbus.send(message)
+        self._module.turn_off(self._channel)
+
+    async def async_added_to_hass(self):
```
This too.

MartinHjelmare commented on homeassistant/components/switch/velbus.py:
```diff
@@ -84,28 +48,27 @@ def should_poll(self):
+        """Add listener for state changes."""
+        await self.hass.async_add_job(self._init_velbus)
+
+    async def async_update(self):
```
This too.

MartinHjelmare commented on homeassistant/components/switch/velbus.py:
```diff
@@ -84,28 +48,27 @@ def should_poll(self):
+        """Update module status."""
+        await self._load_module()
+
+    def _init_velbus(self, callback):
```
This too.

MartinHjelmare commented on homeassistant/components/switch/velbus.py:
```diff
@@ -84,28 +48,27 @@ def should_poll(self):
+        """Initialize Velbus on startup."""
+        self._module.on_status_update(self._channel, callback)
+
+    def _on_update(self, state):
```
This too.

MartinHjelmare commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +27,60 @@
     return True
+
+
+class VelbusEntity:
```
Inherit from `Entity`.

MartinHjelmare commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +27,60 @@
+class VelbusEntity:
+    """Representation of a Velbus entity."""
+
+    def __init__(self, module, channel, hass):
```
Don't pass in `hass`.

MartinHjelmare commented on homeassistant/components/velbus.py:
```diff
@@ -26,18 +27,60 @@
+        """Initialize a Velbus entity."""
+        self._module = module
+        self._channel = channel
+        self.hass = hass
```
Remove this. `hass` will be set on the entity when it has been added to home assistant.

thomasdelaet committed:
integrate review comments (common functionality in VelbusEntity class)

MartinHjelmare reviewed (commented):
Looks great! Two small comments.

MartinHjelmare commented:
Please add a paragraph in the description that describes the breaking change for the release notes.

thomasdelaet committed:
rename DOMAIN import to VELBUS_DOMAIN

thomasdelaet commented:
Added a paragraph with breaking changes

MartinHjelmare reviewed (approved):
:tada:

MartinHjelmare commented:
Can be merged when build passes.

thomasdelaet commented:
Thanks!

thomasdelaet commented:
@MartinHjelmare this is now ready for merging imho

MartinHjelmare commented on requirements_all.txt:
```diff
@@ -424,7 +424,7 @@ hole==0.3.0
 
 # homeassistant.components.frontend
-home-assistant-frontend==20180803.0
+home-assistant-frontend==20180804.0
```
This shouldn't change.

MartinHjelmare commented on requirements_test_all.txt:
```diff
@@ -81,7 +81,7 @@ hbmqtt==0.9.2
 
 # homeassistant.components.frontend
-home-assistant-frontend==20180803.0
+home-assistant-frontend==20180804.0
```
This shouldn't change.

thomasdelaet replied:
The build failed and said that I need to re-run requirements script. This was result. Otherwise build did not work for some reason ...

thomasdelaet replied:
I tried re-running gen_requirements_all.py. If I back out those changes, the build will fail again. What do you suggest I do?

MartinHjelmare reviewed (changes requested):
Maybe run `gen_requirements.py` again? Was it necessary to rebase?

thomasdelaet committed:
revert change created by requirements script

thomasdelaet commented:
@MartinHjelmare : I have made change you requested but now build fails so I am at loss at what to do ...

MartinHjelmare commented:
It seems that the frontend requirement was bumped today without updating the requirements file. Now all builds are failing.

We hold here until we fix this.

MartinHjelmare commented:
Ok, dev branch should be fixed now. Please rebase on latest dev. That should hopefully be enough.

thomasdelaet committed:
regen

thomasdelaet commented:
@MartinHjelmare it's working now.

thomasdelaet commented:
Great! Thanks for your help @MartinHjelmare
